File: Employee.cs
/*
 * Project: ClassComposition
 * Filename: Date.cs
 * Author: J. Hayes
 * Date: Jan. 22, 2020
 * Purpose: To demonstrate how to nest a class within another class, 
 *          i.o., class composition.
 *          The class is the definition of the Employee class. This class
 *          inlcudes the Date class
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ClassComposition
{
    class Employee
    {
        // Date Jan. 27, 2020

        public string FirstName { get; private set;  }
        // Backing variable for FirstName is firstName or _firstName
        public string LastName { get; private set; }
        // Backing variables is lastName or _lastName

        // The following 2 properties are objects of type Date.
        // The relationship between these and the Employee is 
        // a "has-a" relationship. 
        public Date BirthDate { get; private set; }
        public Date HireDate { get; private set; }
        // All the properties are auto-implemented

        // Constructor
        public Employee(string first, string last, Date dOB, Date dOH)
        {
            FirstName = first;
            LastName = last;
            BirthDate = dOB;
            HireDate = dOH;
        }   // end constructor

        // Convert Employee to string format
        public override string ToString()
        {
            return string.Format("{0}, {1}, Hired: {2}, Birth: {3}",
                LastName, FirstName, HireDate, BirthDate);
        }   // end ToString()
    }   // end class
}   // end namespace

File: Employee.cpp
/*
	Filename: Employee.cpp
	Author: J.Hayes
	Date: Nov. 21, 2019
	Purpose: To implement class Employee.
*/

#include <iostream>
#include <string>
#include "Employee.h"
using namespace std;

//	Initialization of the static member variable
int Employee::count = 0;

// Definition of the static member function
int Employee::getCount()
{
	return count;
}	// end getCount()

// Constructor
Employee::Employee(string first, string last)
{
	// firstName = new char[strlen(first) + 1];
	// strcpy(firstName, first);
	firstName = first;

	// lastName = new char[strlen(last) + 1];
	// strcpy(lastName, last);
	lastName = last;

	count++;	// Increment the static member variable

	cout << "Employee constructor for " << firstName << "   " << lastName << "   called." << endl;
}	// end constructor

// Destructor
Employee::~Employee()
{

	cout << "~Employee() called for " << firstName << "   " << lastName << endl;

	// delete[] firstName;
	// delete[] lastName;

	count--;	// Decrement the static member variable
}	// end destructor

string Employee::getFirstName()
{

	return firstName;
}	// end getFirstName()

string Employee::getLastName()
{

	return lastName;
}	// end getLastName()
